Cheaper doc comments

nnethercote opened the pull request:
This PR implements the idea from #60935: represent doc comments more cheaply, rather than converting them into `#[doc="..."]` attribute form. Unlike #60936 (which is about coalescing doc comments to reduce their number), this approach does not have any backwards compatibility concerns, and it eliminates about 80-90% of the current cost of doc comments (as estimated using the numbers in #60930, which eliminated the cost of doc comments entirely by treating them as normal comments).

r? @petrochenkov

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 0b907f7ed55cd4b79ccb380e68fcca0dacc0f0e2 with merge b773dddc5a6434753e502b8beb1868fd476103f8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0b907f7ed55cd4b79ccb380e68fcca0dacc0f0e2","merge_sha":"b773dddc5a6434753e502b8beb1868fd476103f8"} -->

nnethercote commented:
@Manishearth, @nrc: this PR will affect Clippy and rustfmt.

nnethercote commented:
@rust-timer build b773dddc5a6434753e502b8beb1868fd476103f8

rust-timer commented:
Queued b773dddc5a6434753e502b8beb1868fd476103f8 with parent 4a8c5b20c7772bc5342b83d4b0696ea216ef75a7, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4a8c5b20c7772bc5342b83d4b0696ea216ef75a7&end=b773dddc5a6434753e502b8beb1868fd476103f8).

rust-timer commented:
Finished benchmarking try commit b773dddc5a6434753e502b8beb1868fd476103f8, [comparison URL](https://perf.rust-lang.org/compare.html?start=4a8c5b20c7772bc5342b83d4b0696ea216ef75a7&end=b773dddc5a6434753e502b8beb1868fd476103f8).

nnethercote commented:
Perf results are excellent, as expected.

Centril commented:
r? @Centril @bors r+

bors commented:
:pushpin: Commit 42bac153bb9e9baaf30f9aaa2830faf69042ac99 has been approved by `Centril`

<!-- @bors r=Centril 42bac153bb9e9baaf30f9aaa2830faf69042ac99 -->
<!-- homu: {"type":"Approved","sha":"42bac153bb9e9baaf30f9aaa2830faf69042ac99","approver":"Centril"} -->

Centril commented:
@bors rollup=never (perf)

petrochenkov commented:
@bors r-
I really wanted to review this, can it wait for a day or so?

Centril commented:
Sure -- I don't mind.
(I did double check the diff and it all seems good.)

nrc commented:
cc @topecongiro for possible rustfmt breakage

bors commented:
:umbrella: The latest upstream changes (presumably #65804) made this pull request unmergeable. Please resolve the merge conflicts.

nnethercote commented:
I have rebased.

petrochenkov commented on src/libsyntax/ast.rs:
````diff
@@ -2106,20 +2106,54 @@ pub struct AttrItem {
+    ///     }
+    /// )
+    /// ```
+    SugaredDoc(Symbol),
````
```suggestion
    DocComment(Symbol),
```
Could you also mass rename the remaining `SugaredDoc`/`sugared_doc`s into `DocComment`/`doc_comment`s?

nnethercote replied:
I think `SugaredDoc` is better because it clearly only matches sugared doc comments, e.g. `///`. `DocComment` could also mean `#[doc="..."]` comments.

petrochenkov replied:
`#[doc]` is not a comment, it's a regular attribute.
Yes, there are two ways to drive rustdoc - doc comments and `#[doc]` attributes, but that certainly doesn't mean `#[doc]` attributes are comments.
`/// Doc` being a comment and `#[doc]` not being a comment is exactly the point of using the name `DocComment`.

So, `DocComment` literally names the thing it denotes, while `SugaredDoc` doesn't, it's like naming `While` `SugaredLoopWithConditionalBreak` or something.

petrochenkov replied:
So, I disagree and this mixing of doc comments and doc attributes in the frontend is exactly what I'd like to eliminate (or at least minimize).

petrochenkov commented on src/libsyntax/attr/mod.rs:
```diff
@@ -159,29 +166,48 @@ impl Attribute {
+                    None
+                }
+            }
+            AttrKind::SugaredDoc(_) => Some(Ident::with_dummy_span(sym::doc)),
```
```suggestion
            AttrKind::SugaredDoc(_) => Some(Ident::new(sym::doc, self.span)),
```

petrochenkov commented on src/libsyntax/attr/mod.rs:
```diff
@@ -275,17 +301,42 @@ impl AttrItem {
+        match self.kind {
+            AttrKind::Normal(ref item) => item.meta(self.span),
+            AttrKind::SugaredDoc(comment) =>
+                Some(mk_name_value_item_str(Ident::new(sym::doc, DUMMY_SP), comment, DUMMY_SP)),
```
```suggestion
                Some(mk_name_value_item_str(Ident::new(sym::doc, self.span), comment, self.span)),
```

nnethercote replied:
I originally had `self.span` in all the places you suggested, but changed it to `DUMMY_SP` because `self.span` is an approximation at best; there isn't really a correct span possible here. But if span approximation is generally used when desugaring things then I'm fine to continue that trend here.

petrochenkov replied:
Approximation is certainly better than `DUMMY_SP`, all kinds of proc macros and derives span their output to the macro call unless there are better alternatives.
Also, the `SyntaxContext` component of the span is precise in this case.

nnethercote replied:
Thanks for the explanation. I will change this.

petrochenkov commented on src/libsyntax/attr/mod.rs:
```diff
@@ -299,18 +350,25 @@ impl Attribute {
+                })
+            }
+            AttrKind::SugaredDoc(comment) => {
+                Ok(mk_name_value_item_str(Ident::new(sym::doc, DUMMY_SP), comment, DUMMY_SP))
```
```suggestion
                Ok(mk_name_value_item_str(Ident::new(sym::doc, self.span), comment, self.span))
```

petrochenkov commented:
This PR is technically correct, but I disagree with the comment on `AttrKind::SugaredDoc` (https://github.com/rust-lang/rust/pull/65750/files#diff-0de6e63ecfba5c83fb06c9e981f292a6R2122-R2155) entirely and with behavior of `has_name` as well.

Yes, semantically doc comments are equivalent to `#[doc]` at HIR or RustdocIR level, but at syntax/token level they are a separate entity, and doc comments matching `#[doc]` in macros is a huge hack.
So, the new `AttrKind` variant is rather a more faithful source representation in AST than just an optimization.

This is a performance-oriented PR first of all, so I won't bother you with applying the changes that I'd like to make, but rather make one more PR myself once this PR lands.
(Removing the comments on `AttrKind` variants would be nice though, they can just say `A regular attribute *example*` and `A doc comment attribute *example*` for now.)

petrochenkov commented:
@nnethercote 
Regarding the `TokenStream` vs `MetaItemKind` distinction you've mentioned on Discord.

"Meta-items" are an implementation details of built-in attributes specifically, so attributes in general cannot use them.
What they can use is something like this
```rust
struct AttrItem {
    path: Path,
    kind: AttrItemKind,
}

enum AttrItemKind {
    Delimited(TokenStream, DelimToken),
    Word,
    KeyValue(???),
}
```
which is mildly questionable because it's not entirely clear what the `???` in key-value attributes should be and `Word` is arguably a special case of `Delimited`, at least proc macros treat it like that.

The step into the right direction that we can certainly do right now is
```rust
struct AttrItem {
    path: Path,
    delimiter: DelimToken,
    tokens: TokenStream,
}
```
where `delimiter == DelimToken::NoDelim && tokens.is_empty()` means a "word", and `delimiter == DelimToken::NoDelim && !tokens.is_empty()` means a "key-value".

This change is not entirely trivial (or at least was, when I tried to do it as a part of #53293) and is better done in a separate PR.

nnethercote commented:
I was playing around with alternative representations for `AttrItem` as well. I have noticed that the `TokenStream` for the `List` case always has a single `TokenTree::Delimited`, and that the `TokenStream` for the `NameValue` case always starts with a `Eq` token. There is a lot of back-and-forth between token streams and `MetaItemKind`, with trickier cases occurring in `parse_attr_item`, `lower_attr`, and `prepend_attrs`. Finally, `Attribute::parse_meta` and `AttrItem::meta` are doing much the same thing.

It definitely feels like there is a cleaner design waiting to emerge, but I haven't worked out what it is.

petrochenkov commented:
>Finally, `Attribute::parse_meta` and `AttrItem::meta` are doing much the same thing.

The former needs a session and can report errors, the latter cannot report errors and can only return `None` if something is wrong but it doesn't require a session.

All necessary attributes go through `parse_meta()` at least once (in `fn visit_attribute` in `feature_gate/check.rs`), so the errors are certainly reported.
After that `meta()` is called in random places where built-in attributes need to be lowered into their semantic form.

Also, AFAIR, rustdoc doesn't always have the session available when it wants to call `meta()`, so it needs some infrastructural changes if we want to eliminate it.

My plan was to introduce `hir::Attribute` and convert built-in attributes into their semantic form at a single point in HIR lowering (and perhaps in few other places for early processed attributes).
Perhaps then we'll be able to get rid of `MetaItem(Kind)` and `meta()` entirely, or at least turn them into implementation details of AST -> HIR lowering of built-in attributes.

r? @petrochenkov

nnethercote commented:
I have updated the code.

- I replaced the various uses of `DUMMY_SP` with `self.span`, as requested.
- I changed the comments on `AttrKind`, as requested.
- I removed a now-obsolete sentence from the comment on `Attribute::check_name`.
- I left `has_name` unchanged. I can see why you don't like it, but it does
  match the existing behaviour. I am interested to see how you change it in the
  follow-up.
- I left `SugaredDoc` unchanged, for the reasons I explained above.

What do you think, @petrochenkov?

Centril commented on src/libsyntax/ast.rs:
```diff
@@ -2106,20 +2106,28 @@ pub struct AttrItem {
+
+    /// A sugared doc comment attribute (e.g. `/// ...`). Note that this is a
+    /// more compact and thus cheaper representation than unsugared doc
+    /// comments.
```
I feel we could be more positive about this; it's not just an optimization hack but more "correct" from the POV of being syntactically distinct (which is relevant at the level of the concrete syntax tree).

bors commented:
:umbrella: The latest upstream changes (presumably #65885) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
>I left `has_name` unchanged. ... I am interested to see how you change it in the
follow-up.

Make it return `None` for doc comments and avoid using `sym::doc` in most places in general.
(For this PR it makes sense keep the old logic just to minimize changes unrelated to the perf aspect.)

petrochenkov commented on src/libsyntax/ast.rs:
```diff
@@ -2106,16 +2106,30 @@ pub struct AttrItem {
 
+#[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
+pub enum AttrKind {
+    /// A normal attribute. Unsugared doc comments (e.g. `#[doc="..."]`) are
```
These are not comments, as I said above, these are doc attributes.

petrochenkov commented on src/libsyntax/ast.rs:
```diff
@@ -2106,20 +2106,28 @@ pub struct AttrItem {
+    /// represented with this variant.
+    Normal(AttrItem),
+
+    /// A sugared doc comment attribute (e.g. `/// ...`). Note that this is a
```
```suggestion
    /// A doc comment (e.g. `/// ...`).
```

nnethercote commented:
Another update:
- I changed the comments on `AttrKind`, as requested.
- `s/SugaredDoc/DocComment/`, as requested.
- `s/is_sugared_doc()/is_doc_comment()`, as requested.
- `s/mk_sugared_doc_attr()/mk_doc_comment()`.
- `s/with_desugared_doc()/with_doc_comment_markers_stripped()`.
- I updated the commit message.

Note that `DocFragment` uses `SugaredDoc` for doc comments and `RawDoc` for doc
attributes (and "sugared doc"/"raw doc" in comments), so @petrochenkov you might
want to change those as part of your follow-up.

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 9964212e9e74c120d67aca7cae71345d278d76ae has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 9964212e9e74c120d67aca7cae71345d278d76ae -->
<!-- homu: {"type":"Approved","sha":"9964212e9e74c120d67aca7cae71345d278d76ae","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 9964212e9e74c120d67aca7cae71345d278d76ae with merge c154652180cf3be5adb6ad33dc00775e2978c360...
<!-- homu: {"type":"BuildStarted","head_sha":"9964212e9e74c120d67aca7cae71345d278d76ae","merge_sha":"c154652180cf3be5adb6ad33dc00775e2978c360"} -->

rust-highfive commented:
The job `x86_64-gnu-tools` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=12309&view=logs&jobId=e50ab380-190b-535c-8a18-25c988f7577b) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12309/logs/1586), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12309/logs/1586)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-10-30T10:24:32.9334647Z  Caused By: "https://rust-lang.github.io/chalk/doc/chalk_parse/index.html" returned 404 Not Found
2019-10-30T10:24:32.9335129Z  Caused By: "https://rust-lang.github.io/chalk/doc/chalk/index.html" returned 404 Not Found
2019-10-30T10:24:32.9335543Z  Caused By: "https://rust-lang.github.io/chalk/doc/chalki/index.html" returned 404 Not Found
2019-10-30T10:24:32.9336011Z  Caused By: "https://doc.rust-lang.org/nightly/nightly-rustc/syntax/parse/struct.ParseSess.html" returned 404 Not Found
2019-10-30T10:24:32.9336577Z  Caused By: There was an error while fetching "http://www.cs.bgu.ac.il/%7Ehendlerd/papers/p280-hendler.pdf", http://www.cs.bgu.ac.il/%7Ehendlerd/papers/p280-hendler.pdf: timed out
2019-10-30T10:24:32.9395205Z 
2019-10-30T10:24:32.9396669Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rustbook" "linkcheck" "/checkout/src/doc/rustc-guide"
2019-10-30T10:24:32.9397116Z expected success, got: exit code: 101
2019-10-30T10:24:32.9397299Z 
---
2019-10-30T10:30:02.0825484Z [RUSTC-TIMING] cargo_metadata test:false 22.465
2019-10-30T10:30:05.8500472Z error[E0609]: no field `path` on type `&&syntax::ast::Attribute`
2019-10-30T10:30:05.8501001Z   --> src/tools/clippy/clippy_lints/src/utils/attrs.rs:60:35
2019-10-30T10:30:05.8501321Z    |
2019-10-30T10:30:05.8501728Z 60 |         let attr_segments = &attr.path.segments;
2019-10-30T10:30:05.8502545Z    |
2019-10-30T10:30:05.8502946Z    = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:05.8503885Z 
2019-10-30T10:30:06.1397651Z error[E0609]: no field `path` on type `&&syntax::ast::Attribute`
2019-10-30T10:30:06.1397651Z error[E0609]: no field `path` on type `&&syntax::ast::Attribute`
2019-10-30T10:30:06.1398311Z   --> src/tools/clippy/clippy_lints/src/utils/attrs.rs:60:35
2019-10-30T10:30:06.1398659Z    |
2019-10-30T10:30:06.1399067Z 60 |         let attr_segments = &attr.path.segments;
2019-10-30T10:30:06.1400195Z    |
2019-10-30T10:30:06.1400595Z    = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:06.1401487Z 
2019-10-30T10:30:06.1401487Z 
2019-10-30T10:30:07.3756441Z error[E0609]: no field `is_sugared_doc` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.3757633Z     |
2019-10-30T10:30:07.3757633Z     |
2019-10-30T10:30:07.3757927Z 420 |         if attr.is_sugared_doc {
2019-10-30T10:30:07.3758572Z     |
2019-10-30T10:30:07.3758879Z     = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:07.3759345Z 
2019-10-30T10:30:07.3781212Z error[E0609]: no field `tokens` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.3781212Z error[E0609]: no field `tokens` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.3782169Z    --> src/tools/clippy/clippy_lints/src/attrs.rs:424:21
2019-10-30T10:30:07.3782486Z     |
2019-10-30T10:30:07.3783173Z 424 |             if attr.tokens.is_empty() || !is_present_in_source(cx, attr.span) {
2019-10-30T10:30:07.3783813Z     |
2019-10-30T10:30:07.3784132Z     = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:07.3784569Z 
2019-10-30T10:30:07.3784569Z 
2019-10-30T10:30:07.6519326Z error[E0609]: no field `is_sugared_doc` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.6520613Z     |
2019-10-30T10:30:07.6520613Z     |
2019-10-30T10:30:07.6521000Z 420 |         if attr.is_sugared_doc {
2019-10-30T10:30:07.6521819Z     |
2019-10-30T10:30:07.6522215Z     = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:07.6523159Z 
2019-10-30T10:30:07.6523159Z 
2019-10-30T10:30:07.6533759Z error[E0609]: no field `is_sugared_doc` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.6534603Z     |
2019-10-30T10:30:07.6534603Z     |
2019-10-30T10:30:07.6535001Z 250 |         if attr.is_sugared_doc {
2019-10-30T10:30:07.6537297Z     |
2019-10-30T10:30:07.6537997Z     = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:07.6538578Z 
2019-10-30T10:30:07.6581899Z error[E0609]: no field `tokens` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.6581899Z error[E0609]: no field `tokens` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.6582510Z    --> src/tools/clippy/clippy_lints/src/attrs.rs:424:21
2019-10-30T10:30:07.6582835Z     |
2019-10-30T10:30:07.6583286Z 424 |             if attr.tokens.is_empty() || !is_present_in_source(cx, attr.span) {
2019-10-30T10:30:07.6584074Z     |
2019-10-30T10:30:07.6584491Z     = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:07.6585083Z 
2019-10-30T10:30:07.6585083Z 
2019-10-30T10:30:07.9623380Z error[E0609]: no field `is_sugared_doc` on type `&syntax::ast::Attribute`
2019-10-30T10:30:07.9624197Z     |
2019-10-30T10:30:07.9624197Z     |
2019-10-30T10:30:07.9624589Z 250 |         if attr.is_sugared_doc {
2019-10-30T10:30:07.9625363Z     |
2019-10-30T10:30:07.9625767Z     = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:07.9625857Z 
2019-10-30T10:30:08.5877532Z error[E0609]: no field `path` on type `&syntax::ast::Attribute`
2019-10-30T10:30:08.5877532Z error[E0609]: no field `path` on type `&syntax::ast::Attribute`
2019-10-30T10:30:08.5877957Z   --> src/tools/clippy/clippy_lints/src/main_recursion.rs:37:67
2019-10-30T10:30:08.5878253Z    |
2019-10-30T10:30:08.5878618Z 37 |         self.has_no_std_attr = krate.attrs.iter().any(|attr| attr.path == sym::no_std);
2019-10-30T10:30:08.5879302Z    |
2019-10-30T10:30:08.5879621Z    = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:08.5879687Z 
2019-10-30T10:30:08.8483437Z error[E0609]: no field `path` on type `&syntax::ast::Attribute`
2019-10-30T10:30:08.8483437Z error[E0609]: no field `path` on type `&syntax::ast::Attribute`
2019-10-30T10:30:08.8483974Z   --> src/tools/clippy/clippy_lints/src/main_recursion.rs:37:67
2019-10-30T10:30:08.8484332Z    |
2019-10-30T10:30:08.8484776Z 37 |         self.has_no_std_attr = krate.attrs.iter().any(|attr| attr.path == sym::no_std);
2019-10-30T10:30:08.8485644Z    |
2019-10-30T10:30:08.8486033Z    = note: available fields are: `kind`, `id`, `style`, `span`
2019-10-30T10:30:08.8486137Z 
2019-10-30T10:30:10.3342870Z error: aborting due to 5 previous errors
---
2019-10-30T11:02:28.8374815Z   local time: Wed Oct 30 11:02:28 UTC 2019
2019-10-30T11:02:29.3718783Z   network time: Wed, 30 Oct 2019 11:02:29 GMT
2019-10-30T11:02:29.3723064Z == end clock drift check ==
2019-10-30T11:02:32.2423744Z 
2019-10-30T11:02:32.2560218Z ##[error]Bash exited with code '1'.
2019-10-30T11:02:32.2603208Z ##[section]Starting: Upload CPU usage statistics
2019-10-30T11:02:32.2615693Z ==============================================================================
2019-10-30T11:02:32.2615814Z Task         : Bash
2019-10-30T11:02:32.2615919Z Description  : Run a Bash script on macOS, Linux, or Windows
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=12309)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=12309","builder_name":"checks-azure"} -->

nnethercote commented:
```
2019-10-30T10:24:32.9334647Z  Caused By: "https://rust-lang.github.io/chalk/doc/chalk_parse/index.html" returned 404 Not Found
2019-10-30T10:24:32.9335129Z  Caused By: "https://rust-lang.github.io/chalk/doc/chalk/index.html" returned 404 Not Found
2019-10-30T10:24:32.9335543Z  Caused By: "https://rust-lang.github.io/chalk/doc/chalki/index.html" returned 404 Not Found
2019-10-30T10:24:32.9336011Z  Caused By: "https://doc.rust-lang.org/nightly/nightly-rustc/syntax/parse/struct.ParseSess.html" returned 404 Not Found
2019-10-30T10:24:32.9336577Z  Caused By: There was an error while fetching "http://www.cs.bgu.ac.il/%7Ehendlerd/papers/p280-hendler.pdf", http://www.cs.bgu.ac.il/%7Ehendlerd/papers/p280-hendler.pdf: timed out
2019-10-30T10:24:32.9395205Z 
```
Clearly a network error.

And does a successful test run really rely on a successful connection to `www.cs.bgu.ac.il`?

@bors retry

mati865 commented:
@nnethercote it's a non fatal error.
The true reason why it failed is breaking Clippy build since it's ["No Tools Breakage Week"](https://forge.rust-lang.org/#no-tools-breakage-week).
If you want to land it before beta cutoff you'd have to open PR in Clippy repo that fixes the fallout and update Clippy in this PR.

bors commented:
:hourglass: Testing commit 9964212e9e74c120d67aca7cae71345d278d76ae with merge 12c5008749c77ce40adc7037b07433ea582b74f2...
<!-- homu: {"type":"BuildStarted","head_sha":"9964212e9e74c120d67aca7cae71345d278d76ae","merge_sha":"12c5008749c77ce40adc7037b07433ea582b74f2"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=12454)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=12454","builder_name":"checks-azure"} -->

rust-highfive commented:
The job `x86_64-gnu-tools` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=12454&view=logs&jobId=e50ab380-190b-535c-8a18-25c988f7577b) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12454/logs/1521), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12454/logs/1521)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-10-31T20:50:40.9335558Z   local time: Thu Oct 31 20:50:40 UTC 2019
2019-10-31T20:50:41.1967015Z   network time: Thu, 31 Oct 2019 20:50:41 GMT
2019-10-31T20:50:41.1969000Z == end clock drift check ==
2019-10-31T20:50:41.8481173Z 
2019-10-31T20:50:41.8538377Z ##[error]Bash exited with code '1'.
2019-10-31T20:50:41.8574299Z ##[section]Starting: Checkout
2019-10-31T20:50:41.8576965Z ==============================================================================
2019-10-31T20:50:41.8577083Z Task         : Get sources
2019-10-31T20:50:41.8577175Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

petrochenkov commented:
Marking as blocked until the no-breakage week ends.

nnethercote commented:
No Tools Breakage Week is over, or close enough that by the time this gets through the queue it will be.

@bors r=petrochenkov

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #66121

bors commented:
:pushpin: Commit 9964212e9e74c120d67aca7cae71345d278d76ae has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 9964212e9e74c120d67aca7cae71345d278d76ae -->
<!-- homu: {"type":"Approved","sha":"9964212e9e74c120d67aca7cae71345d278d76ae","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 9964212e9e74c120d67aca7cae71345d278d76ae with merge d08f070f1290cecdd37e07a32a3449c253683bec...
<!-- homu: {"type":"BuildStarted","head_sha":"9964212e9e74c120d67aca7cae71345d278d76ae","merge_sha":"d08f070f1290cecdd37e07a32a3449c253683bec"} -->

Centril commented:
Yielding to a r0llup + this can wait a bit more to ensure no toolstate breakage.

@bors retry

bors commented:
:hourglass: Testing commit 9964212e9e74c120d67aca7cae71345d278d76ae with merge bc32c9c57c86f9de80e2d8c99dfb7180c121d74c...
<!-- homu: {"type":"BuildStarted","head_sha":"9964212e9e74c120d67aca7cae71345d278d76ae","merge_sha":"bc32c9c57c86f9de80e2d8c99dfb7180c121d74c"} -->

Centril commented:
Yielding to r0llup. Also, toolstate window is still in effect.

@bors retry

Centril commented:
In the interim, let's @bors p=-1

bors commented:
:umbrella: The latest upstream changes (presumably #66143) made this pull request unmergeable. Please resolve the merge conflicts.

nnethercote commented:
@bors p=0

nnethercote commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit eea6f23a0ed67fd8c6b8e1b02cda3628fee56b2f has been approved by `petrochenkov`

<!-- @bors r=petrochenkov eea6f23a0ed67fd8c6b8e1b02cda3628fee56b2f -->
<!-- homu: {"type":"Approved","sha":"eea6f23a0ed67fd8c6b8e1b02cda3628fee56b2f","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit eea6f23a0ed67fd8c6b8e1b02cda3628fee56b2f with merge caf018714189db0b15f9f803adfcb4572ab7a988...
<!-- homu: {"type":"BuildStarted","head_sha":"eea6f23a0ed67fd8c6b8e1b02cda3628fee56b2f","merge_sha":"caf018714189db0b15f9f803adfcb4572ab7a988"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=12899)
Approved by: petrochenkov
Pushing caf018714189db0b15f9f803adfcb4572ab7a988 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=12899"},"merge_sha":"caf018714189db0b15f9f803adfcb4572ab7a988"} -->

lzutao commented:
The final result: https://perf.rust-lang.org/compare.html?start=38048763e885a3ee139abf39d59a530b16484150&end=caf018714189db0b15f9f803adfcb4572ab7a988&stat=instructions%3Au
